materialize-snowflake/snowflake.go

```diff
@@ -596,93 +609,177 @@ func (d *transactor) Load(it *pm.LoadIterator, loaded func(int, json.RawMessage)
+				// can actually be committed
+				// Important to note that in this case we do not reset the checkpoint for all bindings, but only the ones
+				// that have been committed in this transaction. The reason is that it may be the case that a binding
+				// which has been disabled right after a failed attempt to run its queries, must be able to recover by enabling
+				// the binding and running the queries that are pending for its last transaction.
+				var checkpointClear = checkpoint{Queries: make(map[string][]string), ToDelete: make(map[string][]string)}
+				for _, b := range d.bindings {
+					checkpointClear.Queries[b.target.Identifier] = []string{}
+					checkpointClear.ToDelete[b.target.Identifier] = []string{}
+				}
+				checkpointJSON, err = json.Marshal(checkpointClear)
+				if err != nil {
+					return nil, fmt.Errorf("creating checkpoint clearing json: %w", err)
+				}
+				return &pf.ConnectorState{UpdatedJson: json.RawMessage(checkpointJSON), MergePatch: true}, nil
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			}
+		})
+
+		return &pf.ConnectorState{UpdatedJson: checkpointJSON, MergePatch: true}, commitOp
 	}, nil
 }
 
-func (d *transactor) commit(ctx context.Context) error {
-	var txn, err = d.store.conn.BeginTx(ctx, nil)
-	if err != nil {
-		return fmt.Errorf("store.conn.BeginTx: %w", err)
-	}
-	defer txn.Rollback()
+func renderWithDir(tpl string, dir string) string {
+	return fmt.Sprintf(tpl, dir)
+}
 
-	// First we must validate the fence has not been modified.
-	var fenceUpdate strings.Builder
-	if err := d.templates["updateFence"].Execute(&fenceUpdate, d.store.fence); err != nil {
-		return fmt.Errorf("evaluating fence template: %w", err)
-	} else if _, err = txn.ExecContext(ctx, fenceUpdate.String()); err != nil {
-		err = fmt.Errorf("txn.Exec: %w", err)
+// applyCheckpoint merges data from temporary table to main table
+func (d *transactor) applyCheckpoint(ctx context.Context, cp checkpoint) error {
+	var asyncCtx = sf.WithAsyncMode(ctx)
+	log.Info("store: starting committing changes")
+
+	// Run the queries using AsyncMode, which means that `ExecContext` will not block
+	// until the query is successful, rather we will store the results of these queries
+	// in a map so that we can then call `RowsAffected` on them, blocking until
+	// the queries are actually executed and done
+	var results = make(map[string]stdsql.Result)
+	for table, queries := range cp.Queries {
+		// during recovery we skip queries that belong to tables which do not have a binding anymore
+		// since these tables might be deleted already
+		if !d.hasTableBinding(table) {
+			continue
+		}
 
-		// Give recommendation to user for resolving timeout issues
-		if strings.Contains(err.Error(), "timeout") {
-			return fmt.Errorf("fence.Update: %w  (ensure LOCK_TIMEOUT and STATEMENT_TIMEOUT_IN_SECONDS are at least ten minutes)", err)
+		for _, q := range queries {
+			if result, err := d.store.conn.ExecContext(asyncCtx, q); err != nil {
```

Comment:
Previously we would log out some information about every query as it completed. Could we still do something like that here, with the name of the table coming from the keys of `cp.Queries` and the completion time like it was before?